spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java

```diff
@@ -343,11 +346,22 @@ public void run() {
 						count++;
 					}
 					catch (Exception e) {
-						if (e instanceof RuntimeException) {
-							throw (RuntimeException) e;
+						if (e instanceof MessagingException) {
+							throw (MessagingException) e;
 						}
 						else {
-							throw new MessageHandlingException(new ErrorMessage(e), e);
+							Message<?> failedMessage = null;
+							Object resource = TransactionSynchronizationManager.getResource(getResourceToBind());
+							if (resource instanceof IntegrationResourceHolder) {
+								failedMessage = ((IntegrationResourceHolder) resource).getMessage();
+							}
+							throw new MessagingException(failedMessage, e);
+						}
+					}
+					finally {
+						Object resource = getResourceToBind();
```

Comment:
There is one more in this method - `TransactionSynchronizationManager.getResource(getResourceToBind())`